### PR TITLE
Add users skills logic

### PR DESCRIPTION
**Pull request by BelehovEgor**

### CODE REVIEW

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            migrationBuilder.CreateTable(
+                name: DbSkill.TableName,
+                columns: table => new
+                {
+                    Id = table.Column<Guid>(nullable: false),
+                    SkillName = table.Column<string>(nullable: false)
+                },
+                constraints: table =>
+                {
+                    table.PrimaryKey("PK_Skill", x => x.Id);
```
Comment (borovHokage) [inline src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs:23]: if your table calls "Skills" I think, you should call PrimaryKey "PK_Skills"

Comment (BelehovEgor) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs]: fixed

**Review by borovHokage — APPROVED** [review]

**Review by TIove — APPROVED** [review]

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                },
+                constraints: table =>
+                {
+                    table.PrimaryKey("PK_UserSkills", x => x.Id);
+
+                    table.ForeignKey(
+                        name: "FK_UserSkills_Users",
+                        column: x => x.UserId,
+                        principalTable: DbUser.TableName,
+                        principalColumn: "Id",
```
Comment (djino97) [inline src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs:45]: use nameof(DbUserSkills.Id) and make also for other

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            CreateSkillsTable(migrationBuilder);
+            CreateUsersSkillsTable(migrationBuilder);    
+        }
+
+        protected override void Down(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.DropTable(DbSkill.TableName);
+            migrationBuilder.DropTable(DbUserSkills.TableName);
```
Comment (djino97) [inline src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs:68]: There is the wrong order to drop tables

File: src/UserService.Business/CreateUserCommand.cs [hunk 4]
```diff
@@ -48,12 +49,42 @@ public Guid Execute(UserRequest request)
...
                 request.Login, dbUserCredentials.Salt, request.Password);
 
             return _userRepository.CreateUser(dbUser, dbUserCredentials);
         }
+
+        private void AddUserSkillsToDbUser(DbUser dbUser, UserRequest request)
+        {
+            foreach (var skillName in request.Skills)
+            {
+                var dbSkill = _userRepository.FindSkillByName(skillName);
```
Comment (djino97) [inline src/UserService.Business/CreateUserCommand.cs:67]: add empty line

File: src/UserService.Models.Db/DbUser.cs [hunk 5]
```diff
@@ -1,27 +1,72 @@
...
         public string Status { get; set; }
         public Guid? AvatarFileId { get; set; }
-        [Required]
         public bool IsActive { get; set; }
         public bool IsAdmin { get; set; }
+
         public DbUserCredentials UserCredentials { get; set; }
         public ICollection<DbUserCertificateFile> CertificatesFilesIds { get; set; }
         public ICollection<DbUserAchievement> AchievementsIds { get; set; }
+        public ICollection<DbUserSkills> Skills { get; set; }
```
Comment (djino97) [inline src/UserService.Models.Db/DbUser.cs:26]: rename UserSkills

File: test/UserService.Data.UnitTests/UserRepositoryTests.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        public void ShouldThrowBadRequestExceptionWhenAddedSkillAlreadyInDb()
+        {
+            Assert.Throws<BadRequestException>(() => _repository.CreateSkill(_dbSkillInDb.SkillName));
+        }
+
+        [Test]
+        public void ShouldFindExistSkillByName()
+        {
+            var result = _repository.FindSkillByName(_dbSkillInDb.SkillName);
+            SerializerAssert.AreEqual(_dbSkillInDb, result);
```
Comment (djino97) [inline test/UserService.Data.UnitTests/UserRepositoryTests.cs:77]: add empty line

File: test/UserService.Validation.UnitTests/UserValidatorTests.cs [hunk 7]
```diff
@@ -208,26 +211,67 @@ public void ShouldNotThrowValidationExceptionWhenDataIsValidAndContainsRussianSy
...
+                Login = "admin",
                 Email = "Example@gmail.com",
                 Status = "Example",
                 Password = "Example",
-                IsAdmin = false
+                IsAdmin = false,
+                Skills = new List<string>()
             };
 
             validator.TestValidate(request).ShouldNotHaveAnyValidationErrors();
```
Comment (djino97) [inline test/UserService.Validation.UnitTests/UserValidatorTests.cs:237]: You can use simple `validator.ShouldHaveValidationErrorFor(x => x, object);`

File: test/UserService.Data.UnitTests/UserRepositoryTests.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+            _repository = new UserRepository(_provider);
+        }
+
+        [Test]
+        public void ShouldCreateNewSkillWhenDbHasNotThem()
+        {
+            var nameSkill = "new skill";
+            var result = _repository.CreateSkill(nameSkill);
+            Assert.AreEqual(result, _provider.Skills.FirstOrDefaultAsync(s => s.SkillName == nameSkill).Result.Id);
```
Comment (djino97) [inline test/UserService.Data.UnitTests/UserRepositoryTests.cs:64]: add empty line

File: src/UserService.Data/UserRepository.cs [hunk 9]
```diff
@@ -102,5 +102,25 @@ public IEnumerable<DbUser> GetAllUsers(int skipCount, int takeCount, string user
...
+        public Guid CreateSkill(string name)
+        {
+            var dbSkill = _provider.Skills.FirstOrDefault(s => s.SkillName == name);
+
+            if (dbSkill != null)
+            {
+                throw new BadRequestException();
+            }
+
+            var skill = new DbSkill { Id = Guid.NewGuid(), SkillName = name };
```
Comment (djino97) [inline src/UserService.Data/UserRepository.cs:120]: format code, add empty line.

File: src/UserService.Data/UserRepository.cs [hunk 10]
```diff
@@ -102,5 +102,25 @@ public IEnumerable<DbUser> GetAllUsers(int skipCount, int takeCount, string user
...
+            _provider.Skills.Add(skill);
+            _provider.Save();
+            return skill.Id;
```
Comment (djino97) [inline src/UserService.Data/UserRepository.cs:123]: add empty line

**Review by djino97 — CHANGES_REQUESTED** [review]

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs (hunk 2 above)
Comment (BelehovEgor) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs:45]: fixed

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs (hunk 3 above)
Comment (BelehovEgor) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210317164900_AddDbSkills.cs:68]: table deletions reversed

File: src/UserService.Business/CreateUserCommand.cs (hunk 4 above)
Comment (BelehovEgor) [reply src/UserService.Business/CreateUserCommand.cs:67]: fixed

File: src/UserService.Models.Db/DbUser.cs (hunk 5 above)
Comment (BelehovEgor) [reply src/UserService.Models.Db/DbUser.cs:26]: fixed

File: test/UserService.Data.UnitTests/UserRepositoryTests.cs (hunk 6 above)
Comment (BelehovEgor) [reply test/UserService.Data.UnitTests/UserRepositoryTests.cs:77]: fixed

File: test/UserService.Data.UnitTests/UserRepositoryTests.cs (hunk 8 above)
Comment (BelehovEgor) [reply test/UserService.Data.UnitTests/UserRepositoryTests.cs:64]: fixed

File: src/UserService.Data/UserRepository.cs (hunk 9 above)
Comment (BelehovEgor) [reply src/UserService.Data/UserRepository.cs:120]: fixed

File: src/UserService.Data/UserRepository.cs (hunk 10 above)
Comment (BelehovEgor) [reply src/UserService.Data/UserRepository.cs:123]: fixed

File: test/UserService.Validation.UnitTests/UserValidatorTests.cs (hunk 7 above)
Comment (BelehovEgor) [reply test/UserService.Validation.UnitTests/UserValidatorTests.cs]: I guess now i did what do you want

Comment (BelehovEgor) [reply test/UserService.Validation.UnitTests/UserValidatorTests.cs]: fixed

File: src/UserService.Data/UserRepository.cs [hunk 11]
```diff
@@ -102,5 +102,27 @@ public IEnumerable<DbUser> GetAllUsers(int skipCount, int takeCount, string user
...
+            return _provider.Skills.FirstOrDefault(s => s.SkillName == name);
+        }
+
+        public Guid CreateSkill(string name)
+        {
+            var dbSkill = _provider.Skills.FirstOrDefault(s => s.SkillName == name);
+
+            if (dbSkill != null)
+            {
+                throw new BadRequestException();
```
Comment (djino97) [inline src/UserService.Data/UserRepository.cs:117]: write message in badRequestException

File: src/UserService.Models.Db/DbSkill.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+            builder.HasKey(s => s.Id);
+
+            builder
+                .Property(s => s.SkillName)
+                .IsRequired();
+
+            builder
+                .HasMany(s => s.UserSkills)
+                .WithOne(us => us.Skill)
+                .HasForeignKey(us => us.SkillId);
```
Comment (djino97) [inline src/UserService.Models.Db/DbSkill.cs:38]: not need HasForeignKey

File: src/UserService.Models.Db/DbUser.cs [hunk 13]
```diff
@@ -69,6 +68,11 @@ public void Configure(EntityTypeBuilder<DbUser> builder)
             builder
                 .HasMany(u => u.Connections)
                 .WithOne(conn => conn.User);
+
+            builder
+                .HasMany(u => u.UserSkills)
+                .WithOne(us => us.User)
+                .HasForeignKey(us => us.UserId);
```
Comment (djino97) [inline src/UserService.Models.Db/DbUser.cs:75]: remove HasForeignKey

File: src/UserService.Validation/UserValidator.cs [hunk 14]
```diff
@@ -52,8 +52,13 @@ public UserValidator()
             {
                 RuleForEach(user => user.Connections).ChildRules(c => c.RuleFor(uc => uc.Value).NotEmpty());
             });
-      
 
+            When(user => user.Skills != null, () =>
```
Comment (djino97) [inline src/UserService.Validation/UserValidator.cs:56]: && user.Skills.Any()

File: src/UserService/Controllers/UserController.cs [hunk 15]
```diff
@@ -25,7 +25,7 @@ public User GetUserById([FromServices] IGetUserByIdCommand getUserInfoByIdComman
         }
 
         [HttpPost("createUser")]
```
Comment (djino97) [inline src/UserService/Controllers/UserController.cs:27]: rename createUser => create

**Review by djino97 — CHANGES_REQUESTED** [review]

File: src/UserService.Data/Interfaces/IUserRepository.cs [hunk 16]
```diff
@@ -54,5 +54,19 @@ public interface IUserRepository
...
+        /// </summary>
+        /// <param name="name">Skill name</param>
+        /// <returns></returns>
+        public DbSkill FindSkillByName(string name);
+
+        /// <summary>
+        /// Adds new skill to Database. Returns Id of new DbSkill if it was successful to add
+        /// </summary>
+        /// <param name="name">Skill name</param>
+        /// <returns> Guid of created DbSkill</returns>
```
Comment (djino97) [inline src/UserService.Data/Interfaces/IUserRepository.cs:69]: add dots to the end comments

**Review by djino97 — CHANGES_REQUESTED** [review]

File: src/UserService.Data/UserRepository.cs (hunk 11 above)
Comment (BelehovEgor) [reply src/UserService.Data/UserRepository.cs:117]: fixed

File: src/UserService.Validation/UserValidator.cs (hunk 14 above)
Comment (BelehovEgor) [reply src/UserService.Validation/UserValidator.cs:56]: fixed

File: src/UserService/Controllers/UserController.cs (hunk 15 above)
Comment (BelehovEgor) [reply src/UserService/Controllers/UserController.cs:27]: fixed

File: src/UserService.Data/Interfaces/IUserRepository.cs (hunk 16 above)
Comment (BelehovEgor) [reply src/UserService.Data/Interfaces/IUserRepository.cs:69]: fixed

File: src/UserService.Models.Db/DbSkill.cs (hunk 12 above)
Comment (BelehovEgor) [reply src/UserService.Models.Db/DbSkill.cs:38]: fixed

File: src/UserService.Models.Db/DbUser.cs (hunk 13 above)
Comment (BelehovEgor) [reply src/UserService.Models.Db/DbUser.cs:75]: fixed

**Review by djino97 — APPROVED** [review]

**Pull request merged**